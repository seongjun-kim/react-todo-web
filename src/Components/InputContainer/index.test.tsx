import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import 'jest-styled-components';

import { InputContainer } from './index';
import { ToDoListProvider } from 'Contexts';

describe('<InputContainer />', () => {
  it('renders component correclty', () => {
    const { container } = render(<InputContainer />);

    const input = screen.getByPlaceholderText('Enter what to do...');
    expect(input).toBeInTheDocument();
    const button = screen.getByText('Add');
    expect(button).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('clear data after adding data', () => {
    render(<InputContainer />);

    const input = screen.getByPlaceholderText('Enter what to do...') as HTMLInputElement;
    const button = screen.getByText('Add');

    expect(input.value).toBe('');
    const sampleInput = 'ToDo 1';
    fireEvent.change(input, { target: { value: sampleInput } });
    expect(input.value).toBe(sampleInput);
    fireEvent.click(button);
    expect(input.value).toBe('');
  });

  it('adds input data to localStorage via Context', () => {
    render(
      <ToDoListProvider>
        <InputContainer />
      </ToDoListProvider>,
    );

    const input = screen.getByPlaceholderText('Enter what to do...') as HTMLInputElement;
    const addButton = screen.getByText('Add');
    const sampleInput = 'New ToDo';

    expect(localStorage.getItem('ToDoList')).toBeNull();
    fireEvent.change(input, { target: { value: sampleInput } });
    fireEvent.click(addButton);
    // expect(localStorage.getItem('ToDoList')).toBe(`["${sampleInput}"]`); // 저장된 데이터 형태로 확인
    expect(JSON.parse(localStorage.getItem('ToDoList') as string).length).toBe(1); // 저장된 데이터 갯수로 확인
  });

  it('calls the onAdd function when the user clicks add button', () => {
    const handleClick = jest.fn();
    render(<InputContainer onAdd={handleClick} />);

    const input = screen.getByPlaceholderText('Enter what to do...');
    const button = screen.getByText('Add');
    expect(handleClick).toHaveBeenCalledTimes(0);

    fireEvent.click(button);
    expect(handleClick).toHaveBeenCalledTimes(0);

    fireEvent.change(input, { target: { value: 'ToDo 1' } });
    fireEvent.click(button);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});
