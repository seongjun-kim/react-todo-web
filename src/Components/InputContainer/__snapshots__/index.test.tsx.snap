// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputContainer /> renders component correclty 1`] = `
.c2 {
  text-align: center;
  background-color: #304ffe;
  padding: 10px 20px;
  border-radius: 8px;
  cursor: pointer;
}

.c2:hover {
  background-color: #1e40ff;
}

.c2:active {
  box-shadow: inset 5px 5px 10px rgba(0,0,0,0.2);
}

.c3 {
  color: #fff;
  font-size: 16px;
}

.c1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-right: 10px;
  font-size: 16px;
  padding: 10px 10px;
  border-radius: 8px;
  border: 1px solid #bdbdbd;
  outline: none;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1"
      placeholder="Enter what to do..."
      value=""
    />
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        Add
      </div>
    </div>
  </div>
</div>
`;
